### PR TITLE
Report compiler errors in machine readable format

### PR DESCRIPTION
**Pull request by cmaglie**
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [X] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [X] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)
- [ ] `configuration.schema.json` updated if new parameters are added.

## What kind of change does this PR introduce?

Adds parsing of the compiler errors and warning to present them in a machine-readable way if JSON output is selected or if the `Compile` gRPC call is used.

## What is the current behavior?

The compiler output is presented as a big blob of text.

## What is the new behavior?

The compiler output is dissected into a structured format so clients can easily consume it.

```
$ arduino-cli compile -u -b arduino:avr:leonardo ~/Arduino/Blink/ -v --format json |jq .builder_result.diagnostics
[
  {
    "column": 16,
    "context": [
      {
        "file": "/home/megabug/Workspace/sketchbook-cores-beta/libraries/Audio/src/Audio.h",
        "line": 16,
        "message": "included from here"
      },
      {
        "file": "/home/megabug/Arduino/Blink/Blink.ino",
        "line": 1,
        "message": "included from here"
      }
    ],
    "file": "/home/megabug/Workspace/sketchbook-cores-beta/libraries/Audio/src/DAC.h",
    "line": 21,
    "message": "expected ')' before '*' token\n  DACClass(Dacc *_dac, uint32_t _dacId, IRQn_Type _isrId) :\n                ^",
    "severity": "ERROR"
  },
  {
    "column": 2,
    "context": [
      {
        "file": "/home/megabug/Workspace/sketchbook-cores-beta/libraries/Audio/src/Audio.h",
        "line": 16,
        "message": "included from here"
      },
      {
        "file": "/home/megabug/Arduino/Blink/Blink.ino",
        "line": 1,
        "message": "included from here"
      }
    ],
    "file": "/home/megabug/Workspace/sketchbook-cores-beta/libraries/Audio/src/DAC.h",
    "line": 35,
    "message": "'Dacc' does not name a type\n  Dacc *dac;\n  ^~~~",
    "severity": "ERROR"
  },
  ...
```

## Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?

No

## Other information

Fix https://github.com/arduino/arduino-cli/issues/1121

### CODE REVIEW

**Comment by cmaglie** [comment]
TODO: test this with more compilers...

File: internal/builder/diagnostics/compiler_detection.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		out := &bytes.Buffer{}
+		cmd.RedirectStdoutTo(out)
+		cmd.Run()
+		return splitLines(out.Bytes())
+	}
+	return nil
+}
+
+// DetectCompilerFromCommandLine tries to detect a compiler from a given command line.
+// If probleCompiler is true, the compiler may be executed with different flags to
```
Comment (alessio-perugini) [inline internal/builder/diagnostics/compiler_detection.go:47]:
```suggestion
// If probeCompiler is true, the compiler may be executed with different flags to
```
typo

File: internal/builder/diagnostics/parser_test.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	require.NotNil(t, detectedCompiler)
+	diags, err := ParseCompilerOutput(detectedCompiler, data)
+	require.NoError(t, err)
+
+	// Check if the parsed data match the expected output
+	output, err := json.MarshalIndent(diags, "", "  ")
+	require.NoError(t, err)
+	golden, err := paths.New("testdata", "compiler_outputs", testFile+".json").ReadFile()
+	require.NoError(t, err)
+	// fmt.Println(string(output))
```
Comment (alessio-perugini) [inline internal/builder/diagnostics/parser_test.go:55]:
```suggestion
```
I guess some debbugging leftovers :nerd_face:

**Review by alessio-perugini — DISMISSED** [review]
LGTM. The only think that I'm concerned about is that starting from now on this is something new that we need to maintain so here are some ideas for future PRs:
- Add more tests for different versions of compilers we're using.
- Adding fuzz testing in the parser might be helpful?

**Comment by cmaglie** [comment]
> Add more tests for different versions of compilers we're using.

This will surely help. Adding a test is a tedious process tho...

> Adding fuzz testing in the parser might be helpful?

Interesting, BTW I think that the fuzzer may not help to improve the parsing correctness because:
* we have no way to check if the input produced by the fuzzer is a plausible gcc output
* we have no way to check if the parser output is correct against the given fuzzer input

Anyway, it may help to test the parser's resistance to crazy inputs...

**Review by alessio-perugini — CHANGES_REQUESTED** [review]
I'd like to have an integration test that checks for the new `diagnostic` field.
If you want you can either:
- Expand the already existing integration test `internal/integrationtest/compile_3/compile_test.go|102 col 6| func TestCompilerErrOutput(t *testing.T) {` 
- Create a new one

Example:
```go
TestCompilerErrOutput(t *testing.T) {
	...
	// prepare sketch
	sketch, err := paths.New("testdata", "blink_with_wrong_cpp").Abs()
	require.NoError(t, err)

	// Run compile and catch err stream
	out, _, err := cli.Run("compile", "-b", "arduino:avr:uno", "--format", "json", sketch.String())
	...
	requirejson.Query(t, out, `.diagnostics | length > 0`, `true`)
```

**Review by alessio-perugini — APPROVED** [review]

**Review by alessio-perugini — APPROVED** [review]

**Pull request merged**